File: src/main/java/com/kinvey/sample/statusshare/component/CommentAdapter.java
/** 
 * Copyright (c) 2013, Kinvey, Inc. All rights reserved.
 *
 * This software contains valuable confidential and proprietary information of
 * KINVEY, INC and is subject to applicable licensing agreements.
 * Unauthorized reproduction, transmission or distribution of this file and its
 * contents is a violation of applicable laws.
 * 
 */
package com.kinvey.sample.statusshare.component;

import android.content.Context;
import android.graphics.Typeface;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.TextView;
import com.google.api.client.json.GenericJson;
import com.google.api.client.util.ArrayMap;
import com.kinvey.sample.statusshare.MainActivity;
import com.kinvey.sample.statusshare.R;


import java.text.ParsePosition;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Date;
import java.util.List;
import java.util.Locale;

public class CommentAdapter extends ArrayAdapter<GenericJson> {

    private LayoutInflater mInflater;
    private Typeface roboto;

    public CommentAdapter(Context context, List<GenericJson> objects,
                          LayoutInflater inf) {
        // NOTE: I pass an arbitrary textViewResourceID to the super
        // constructor-- Below I override
        // getView(...), which causes the underlying adapter to ignore this
        // field anyways, it is just needed in the constructor.
        super(context, 0, objects);
        this.mInflater = inf;
        roboto = Typeface.createFromAsset(context.getAssets(), "Roboto-Thin.ttf");

    }


    @Override
    public View getView(int position, View convertView, ViewGroup parent) {
        UpdateViewHolder holder = null;

        TextView comment = null;
        TextView author = null;
        TextView when = null;

        GenericJson rowData = getItem(position);

        if (null == convertView) {
            convertView = mInflater.inflate(R.layout.row_comment, null);
            holder = new UpdateViewHolder(convertView);
            convertView.setTag(holder);
        }
        holder = (UpdateViewHolder) convertView.getTag();

        if (rowData.get("text") != null){
            comment = holder.getBlurb();
            comment.setText(rowData.get("text").toString());
        }

        if (rowData.get("author") != null){
            author = holder.getAuthor();
            author.setText(rowData.get("author").toString());

        }



        ParsePosition pp = new ParsePosition(0);
        Date date = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS", Locale.US).parse(((ArrayMap)rowData.get("_kmd")).get("ect").toString(), pp);
        String since = MainActivity.getSince(date, Calendar.getInstance());
        if (since != null){
            when = holder.getWhen();
            when.setText(since);

        }

        return convertView;
    }


    /**
     * This pattern is used as an optimization for Android ListViews.
     * <p/>
     * Since every row uses the same layout, the View object itself can be
     * recycled, only the data/content of the row has to be updated.
     * <p/>
     * This allows for Android to only inflate enough Row Views to fit on
     * screen, and then they are recycled. This allows us to avoid creating
     * a new view for every single row, which can have a negative effect on
     * performance (especially with large lists on large screen devices).
     */
    private class UpdateViewHolder {
        private View rowView;

        private TextView blurb = null;
        private TextView author = null;
        private TextView when = null;

        public UpdateViewHolder(View row) {
            rowView = row;
        }

        public TextView getWhen() {
            if (null == when) {
                when = (TextView) rowView.findViewById(R.id.row_comment_time);
            }
            when.setTypeface(roboto);
            return when;
        }

        public TextView getAuthor() {
            if (null == author) {
                author = (TextView) rowView.findViewById(R.id.row_comment_author);
            }
            author.setTypeface(roboto);
            return author;
        }

        public TextView getBlurb() {
            if (null == blurb) {
                blurb = (TextView) rowView.findViewById(R.id.row_comment_text);
            }
            blurb.setTypeface(roboto);
            return blurb;
        }

    }
}

File: app/src/main/java/com/kinvey/sample/statusshare/ui/component/CommentAdapter.kt
/**
 * Copyright (c) 2019, Kinvey, Inc. All rights reserved.
 *
 * This software contains valuable confidential and proprietary information of
 * KINVEY, INC and is subject to applicable licensing agreements.
 * Unauthorized reproduction, transmission or distribution of this file and its
 * contents is a violation of applicable laws.
 *
 */
package com.kinvey.sample.statusshare.ui.component

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.TextView
import com.kinvey.sample.statusshare.utils.Constants.ECT_FIELD_NAME
import com.kinvey.sample.statusshare.utils.Constants.SERVER_DATE_FMT
import com.kinvey.sample.statusshare.ui.MainActivity
import com.kinvey.sample.statusshare.R.id
import com.kinvey.sample.statusshare.R.layout
import com.kinvey.sample.statusshare.model.CommentEntity
import org.json.JSONException
import org.json.JSONObject
import timber.log.Timber
import java.text.ParsePosition
import java.text.SimpleDateFormat
import java.util.*

class CommentAdapter // NOTE: I pass an arbitrary textViewResourceID to the super
// constructor-- Below I override
// getView(...), which causes the underlying adapter to ignore this
// field anyways, it is just needed in the constructor.
(context: Context?, objects: List<CommentEntity?>, private val mInflater: LayoutInflater) 
    : ArrayAdapter<CommentEntity?>(context!!, 0, objects) {
    
    private val format = SimpleDateFormat(SERVER_DATE_FMT, Locale.US)
    private val zeroPp = ParsePosition(0)
    
    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        var convertView = convertView
        var holder: UpdateViewHolder? = null
        var commentText: TextView? = null
        var authorText: TextView? = null
        var whenText: TextView? = null
        val rowData = getItem(position)
        if (null == convertView) {
            convertView = mInflater.inflate(layout.row_comment, null)
            holder = UpdateViewHolder(convertView)
            convertView.tag = holder
        }
        holder = convertView?.tag as UpdateViewHolder
        if (rowData?.text != null) {
            commentText = holder.blurb
            commentText?.text = rowData.text
        }
        if (rowData?.author != null) {
            authorText = holder?.author
            authorText?.text = rowData.author
        }
        val dateString = rowData?.meta.toString()
        try {
            val jsonObj = JSONObject(dateString)
            val ectDateStr: String = jsonObj.getString(ECT_FIELD_NAME)
            val date: Date? = format.parse(ectDateStr, zeroPp)
            val since: String? = MainActivity.getSince(date, Calendar.getInstance())
            if (since != null) {
                whenText = holder?.whenText
                whenText?.text = since
            }
        } catch (e: JSONException) {
            Timber.d("comment JSONException")
        }
        return convertView
    }

    /**
     * This pattern is used as an optimization for Android ListViews.
     *
     *
     * Since every row uses the same layout, the View object itself can be
     * recycled, only the data/content of the row has to be updated.
     *
     *
     * This allows for Android to only inflate enough Row Views to fit on
     * screen, and then they are recycled. This allows us to avoid creating
     * a new view for every single row, which can have a negative effect on
     * performance (especially with large lists on large screen devices).
     */
    private inner class UpdateViewHolder(private val rowView: View) {
        var blurb: TextView? = null
            get() {
                if (null == field) {
                    field = rowView.findViewById<View>(id.row_comment_text) as TextView
                }
                return field
            }
            private set
        var author: TextView? = null
            get() {
                if (null == field) {
                    field = rowView.findViewById<View>(id.row_comment_author) as TextView
                }
                return field
            }
            private set
        var whenText: TextView? = null
            get() {
                if (null == field) {
                    field = rowView.findViewById<View>(id.row_comment_time) as TextView
                }
                return field
            }
            private set
    }
}